File: examples/taskstore/elixir/lib/wire.ex
# Copyright 2026 Chengxi Luo
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

defmodule Wire do
  @moduledoc """
  Byte-level primitives. Go owns the layout these reproduce; the conventions
  are documented at the top of go/api/wire.go.

  Elixir has no cursor: binary pattern matching takes the head off and hands
  back the tail, so every reader here returns `{value, rest}`. Same layout as
  everywhere else — the byte order and the widths are the contract, not the
  shape of the code that walks them.
  """

  @priorities ["low", "normal", "high"]
  @error_codes ["not_found", "invalid", "conflict"]

  @doc "Declaration order of the `enum<low, normal, high>` in cases/task.yaml."
  def priorities, do: @priorities

  @doc "Declaration order of the enum in cases/api_error.yaml."
  def error_codes, do: @error_codes

  @doc "A u32 byte length followed by the bytes themselves."
  def str(s), do: <<byte_size(s)::little-32, s::binary>>

  def take_str(<<n::little-32, s::binary-size(n), rest::binary>>), do: {s, rest}

  def bool(true), do: <<1>>
  def bool(false), do: <<0>>

  @doc """
  An enum as the ordinal of its position in `variants`.

  The ordinal is this codec's own byte-layout choice — an enum travels through
  serify as its *name* — so the list has to match the case file's declaration
  order.
  """
  def enum(variants, name) do
    case Enum.find_index(variants, &(&1 == name)) do
      nil -> raise ArgumentError, "#{inspect(name)} is not one of #{inspect(variants)}"
      ord -> <<ord::8>>
    end
  end

  def take_enum(variants, <<ord::8, rest::binary>>) do
    case Enum.at(variants, ord) do
      nil -> raise ArgumentError, "enum ordinal #{ord} out of range"
      name -> {name, rest}
    end
  end

  def tags(tags), do: <<length(tags)::little-32>> <> Enum.map_join(tags, &str/1)

  def take_tags(<<n::little-32, rest::binary>>), do: take_n(rest, n, &take_str/1)

  def optional_i64(nil), do: <<0>>
  def optional_i64(v), do: <<1, v::signed-little-64>>

  def take_optional_i64(<<0, rest::binary>>), do: {nil, rest}
  def take_optional_i64(<<1, v::signed-little-64, rest::binary>>), do: {v, rest}

  @doc "Take `n` items with `fun`, returning them in order plus the remaining bytes."
  def take_n(data, n, fun) do
    Enum.reduce(1..n//1, {[], data}, fn _, {acc, rest} ->
      {item, rest} = fun.(rest)
      {[item | acc], rest}
    end)
    |> then(fn {acc, rest} -> {Enum.reverse(acc), rest} end)
  end
end

File: examples/taskstore/rust/src/wire.rs
// Copyright 2026 Chengxi Luo
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Byte-level primitives. Go owns the layout these reproduce; the conventions
//! are documented at the top of go/api/wire.go.

/// A cursor over one message's bytes.
///
/// Go threads the remaining slice through every read and Python keeps an
/// offset; Rust is happiest with a borrow plus a position. The byte order and
/// the widths are the contract — the shape of the code that walks them is not.
pub struct Reader<'a> {
    data: &'a [u8],
    pos: usize,
}

impl<'a> Reader<'a> {
    pub fn new(data: &'a [u8]) -> Self {
        Self { data, pos: 0 }
    }

    fn take(&mut self, n: usize) -> Result<&'a [u8], String> {
        if self.pos + n > self.data.len() {
            return Err(format!(
                "truncated: want {n} bytes at offset {}, have {}",
                self.pos,
                self.data.len() - self.pos
            ));
        }
        let out = &self.data[self.pos..self.pos + n];
        self.pos += n;
        Ok(out)
    }

    pub fn u8(&mut self) -> Result<u8, String> {
        Ok(self.take(1)?[0])
    }

    pub fn u32(&mut self) -> Result<u32, String> {
        Ok(u32::from_le_bytes(self.take(4)?.try_into().unwrap()))
    }

    pub fn u64(&mut self) -> Result<u64, String> {
        Ok(u64::from_le_bytes(self.take(8)?.try_into().unwrap()))
    }

    pub fn i64(&mut self) -> Result<i64, String> {
        Ok(i64::from_le_bytes(self.take(8)?.try_into().unwrap()))
    }

    pub fn bool(&mut self) -> Result<bool, String> {
        Ok(self.u8()? != 0)
    }

    pub fn string(&mut self) -> Result<String, String> {
        let n = self.u32()? as usize;
        String::from_utf8(self.take(n)?.to_vec()).map_err(|e| e.to_string())
    }

    /// An enum arrives as an ordinal into a fixed list, so a value outside the
    /// declared variants has no encoding and cannot be read back.
    pub fn enum_of(&mut self, variants: &[&str]) -> Result<String, String> {
        let ord = self.u8()? as usize;
        variants
            .get(ord)
            .map(|s| s.to_string())
            .ok_or_else(|| format!("enum ordinal {ord} out of range for {variants:?}"))
    }

    pub fn tags(&mut self) -> Result<Vec<String>, String> {
        let n = self.u32()?;
        (0..n).map(|_| self.string()).collect()
    }

    pub fn optional_i64(&mut self) -> Result<Option<i64>, String> {
        if self.u8()? == 0 {
            return Ok(None);
        }
        self.i64().map(Some)
    }
}

pub fn put_str(buf: &mut Vec<u8>, s: &str) {
    buf.extend_from_slice(&(s.len() as u32).to_le_bytes());
    buf.extend_from_slice(s.as_bytes());
}

pub fn put_enum(buf: &mut Vec<u8>, variants: &[&str], name: &str) -> Result<(), String> {
    let ord = variants
        .iter()
        .position(|v| *v == name)
        .ok_or_else(|| format!("{name:?} is not one of {variants:?}"))?;
    buf.push(ord as u8);
    Ok(())
}

pub fn put_tags(buf: &mut Vec<u8>, tags: &[String]) {
    buf.extend_from_slice(&(tags.len() as u32).to_le_bytes());
    for t in tags {
        put_str(buf, t);
    }
}

pub fn put_optional_i64(buf: &mut Vec<u8>, v: Option<i64>) {
    match v {
        None => buf.push(0),
        Some(n) => {
            buf.push(1);
            buf.extend_from_slice(&n.to_le_bytes());
        }
    }
}
